apps/aecore/src/aec_consensus_hc.erl

```diff
@@ -267,12 +267,20 @@ state_pre_transform_node(Type, Height, PrevNode, Trees) ->
     {ok, PrevHash} = aec_headers:hash_header(PrevHeader),
     {TxEnv0, _} = aetx_env:tx_env_and_trees_from_hash(aetx_transaction, PrevHash),
     TxEnv = aetx_env:set_height(TxEnv0, Height),
-    {ok, #{epoch := Epoch} = EpochInfo} = aec_chain_hc:epoch_info({TxEnv, Trees}),
+    {ok, #{epoch := Epoch, first := EpochFirst, last := EpochLast} = EpochInfo} =
+        aec_chain_hc:epoch_info({TxEnv, Trees}),
     {ok, Leader} = leader_for_height(Height, {TxEnv, Trees}),
     case Type of
         key ->
-            case Height == maps:get(last, EpochInfo) of
-                true ->
+            if Height =:= EpochFirst ->
+                    %% cache the current epoch start time
+                    EpochStartTime = aetx_env:time_in_msecs(TxEnv),
+                    cache_child_epoch_info(Epoch, Height, EpochStartTime);
+               true ->
+                    ok
+            end,
+            %% note that EpochFirst and EpochLast could be the same, not exclusive
+            if Height =:= EpochLast ->
```

Comment:
One if should be enough?

Reply:
Both cases need to run independently in case they overlap.

Reply:
Probably not, but the code is independent and should not be conflated into a single construct, IMO.